[ICE]: index out of bounds: the len is 0 but the index is 0  (`type_at(0)`)

<!--
[31mICE[0m: Rustc ./a.rs '-Znext-solver=globally -ooutputfile -Zdump-mir-dir=dir' 'thread 'rustc' ($TID) panicked at compiler/rustc_type_ir/src/term_kind.rs:252:19: 'index out of bounds: the len is 0 but the index is 0'', 'thread 'rustc' ($TID) panicked at compiler/rustc_type_ir/src/term_kind.rs:252:19: 'index out of bounds: the len is 0 but the index is 0''
File: /tmp/im/15/report/a.rs
-->

auto-reduced (treereduce-rust):
````rust
//@compile-flags: -Znext-solver=globally
#![feature(inherent_associated_types)]

struct Foo<T>(T);

impl Foo<Vec<[u32]>> {
    type Assoc = u32;
}

type Tait = impl Sized;

#[define_opaque(Tait)]
fn bar() {
    let x: Foo<Tait>::Assoc = 42;
}

````

original:
````rust
//@compile-flags: -Znext-solver=globally
#![feature(inherent_associated_types)]
#![feature(type_alias_impl_trait)]

struct Foo<T>(T);

impl Foo<Vec<[u32]>> {
    type Assoc = u32;
}

type Tait = impl Sized;

#[define_opaque(Tait)]
fn bar() {
    let x: Foo<Tait>::Assoc = 42;
}

fn main() {}

````

Version information
````
rustc 1.98.0-nightly (fd07dbfc9 2026-06-28)
binary: rustc
commit-hash: fd07dbfc91b7b6b3fa76d957e130c51e654131ee
commit-date: 2026-06-28
host: x86_64-unknown-linux-gnu
release: 1.98.0-nightly
LLVM version: 22.1.8

````

Possibly related line of code:
https://github.com/rust-lang/rust/blob/fd07dbfc91b7b6b3fa76d957e130c51e654131ee/compiler/rustc_type_ir/src/term_kind.rs#L246-L258

Command:
`/home/matthias/.rustup/toolchains/master/bin/rustc -Znext-solver=globally`

<details><summary><strong>Program output</strong></summary>
<p>

```
error[E0658]: use of unstable library feature `type_alias_impl_trait`: `type_alias_impl_trait` has open design concerns
  --> /tmp/icemaker_global_tempdir.2c2tnFStzD5n/rustc_testrunner_tmpdir_reporting.FAOLOlpFgGLp/mvce.rs:11:3
   |
11 | #[define_opaque(Tait)]
   |   ^^^^^^^^^^^^^
   |
   = note: see issue #63063 <https://github.com/rust-lang/rust/issues/63063> for more information
   = help: add `#![feature(type_alias_impl_trait)]` to the crate attributes to enable
   = note: this compiler was built on 2026-06-28; consider upgrading it if it is out of date

error[E0658]: `impl Trait` in type aliases is unstable
 --> /tmp/icemaker_global_tempdir.2c2tnFStzD5n/rustc_testrunner_tmpdir_reporting.FAOLOlpFgGLp/mvce.rs:9:13
  |
9 | type Tait = impl Sized;
  |             ^^^^^^^^^^
  |
  = note: see issue #63063 <https://github.com/rust-lang/rust/issues/63063> for more information
  = help: add `#![feature(type_alias_impl_trait)]` to the crate attributes to enable
  = note: this compiler was built on 2026-06-28; consider upgrading it if it is out of date

warning: the feature `inherent_associated_types` is incomplete and may not be safe to use and/or cause compiler crashes
 --> /tmp/icemaker_global_tempdir.2c2tnFStzD5n/rustc_testrunner_tmpdir_reporting.FAOLOlpFgGLp/mvce.rs:1:12
  |
1 | #![feature(inherent_associated_types)]
  |            ^^^^^^^^^^^^^^^^^^^^^^^^^
  |
  = note: see issue #8995 <https://github.com/rust-lang/rust/issues/8995> for more information
  = note: `#[warn(incomplete_features)]` on by default

error[E0601]: `main` function not found in crate `mvce`
  --> /tmp/icemaker_global_tempdir.2c2tnFStzD5n/rustc_testrunner_tmpdir_reporting.FAOLOlpFgGLp/mvce.rs:14:2
   |
14 | }
   |  ^ consider adding a `main` function to `/tmp/icemaker_global_tempdir.2c2tnFStzD5n/rustc_testrunner_tmpdir_reporting.FAOLOlpFgGLp/mvce.rs`

error[E0277]: the size for values of type `[u32]` cannot be known at compilation time
   --> /tmp/icemaker_global_tempdir.2c2tnFStzD5n/rustc_testrunner_tmpdir_reporting.FAOLOlpFgGLp/mvce.rs:5:10
    |
  5 | impl Foo<Vec<[u32]>> {
    |          ^^^^^^^^^^ doesn't have a size known at compile-time
    |
    = help: the trait `Sized` is not implemented for `[u32]`
note: required by an implicit `Sized` bound in `Vec`
   --> /home/matthias/.rustup/toolchains/master/lib/rustlib/src/rust/library/alloc/src/vec/mod.rs:436:16
    |
436 | pub struct Vec<T, #[unstable(feature = "allocator_api", issue = "32838")] A: Allocator = Global> {
    |                ^ required by the implicit `Sized` requirement on this type parameter in `Vec`


thread 'rustc' (1878989) panicked at /rustc-dev/fd07dbfc91b7b6b3fa76d957e130c51e654131ee/compiler/rustc_type_ir/src/term_kind.rs:252:19:
index out of bounds: the len is 0 but the index is 0
stack backtrace:
   0:     0x7f1dcfad3d09 - <<std[fd5358557e4939aa]::sys::backtrace::BacktraceLock>::print::DisplayBacktrace as core[824893b31d0d62cd]::fmt::Display>::fmt
   1:     0x7f1dd020f288 - core[824893b31d0d62cd]::fmt::write
   2:     0x7f1dcfaea376 - <std[fd5358557e4939aa]::sys::stdio::unix::Stderr as std[fd5358557e4939aa]::io::Write>::write_fmt
   3:     0x7f1dcfaa8e6e - std[fd5358557e4939aa]::panicking::default_hook::{closure#0}
   4:     0x7f1dcfac6883 - std[fd5358557e4939aa]::panicking::default_hook
   5:     0x7f1dceab63a1 - std[fd5358557e4939aa]::panicking::update_hook::<alloc[b6d51ad5583e1cd2]::boxed::Box<rustc_driver_impl[5bc58d79d7fe6386]::install_ice_hook::{closure#1}>>::{closure#0}
   6:     0x7f1dcfac6cd2 - std[fd5358557e4939aa]::panicking::panic_with_hook
   7:     0x7f1dcfaa8f32 - std[fd5358557e4939aa]::panicking::panic_handler::{closure#0}
   8:     0x7f1dcfa9fbf9 - std[fd5358557e4939aa]::sys::backtrace::__rust_end_short_backtrace::<std[fd5358557e4939aa]::panicking::panic_handler::{closure#0}, !>
   9:     0x7f1dcfaaa99d - __rustc[3637667c58aa7985]::rust_begin_unwind
  10:     0x7f1dccbca1dc - core[824893b31d0d62cd]::panicking::panic_fmt
  11:     0x7f1dcc9ab3a9 - core[824893b31d0d62cd]::panicking::panic_bounds_check
  12:     0x7f1dd22ad4cc - <&rustc_middle[eddce6fccf027369]::ty::list::RawList<(), rustc_middle[eddce6fccf027369]::ty::generic_args::GenericArg> as rustc_type_ir[a06be78d10bf2c46]::inherent::GenericArgs<rustc_middle[eddce6fccf027369]::ty::context::TyCtxt>>::type_at.cold
  13:     0x7f1dd0722c2d - <dyn rustc_hir_analysis[4b6d32d55a266fd3]::hir_ty_lowering::HirTyLowerer>::lower_type_relative_path::{closure#0}
  14:     0x7f1dd072b486 - <dyn rustc_hir_analysis[4b6d32d55a266fd3]::hir_ty_lowering::HirTyLowerer>::lower_type_relative_ty_path::{closure#0}
  15:     0x7f1dd0730646 - <dyn rustc_hir_analysis[4b6d32d55a266fd3]::hir_ty_lowering::HirTyLowerer>::lower_ty
  16:     0x7f1dd04596d3 - <rustc_hir_typeck[e1f0f3104d94275e]::gather_locals::GatherLocalsVisitor>::declare
  17:     0x7f1dd0455a8a - <rustc_hir_typeck[e1f0f3104d94275e]::fn_ctxt::FnCtxt>::check_expr_block
  18:     0x7f1dd04831eb - <rustc_hir_typeck[e1f0f3104d94275e]::fn_ctxt::FnCtxt>::check_expr_with_expectation_and_args
  19:     0x7f1dd0d81539 - rustc_hir_typeck[e1f0f3104d94275e]::check::check_fn
  20:     0x7f1dd143c79c - rustc_hir_typeck[e1f0f3104d94275e]::typeck_with_inspect::{closure#0}
  21:     0x7f1dd143b130 - rustc_query_impl[363d8420d406e095]::query_impl::typeck_root::invoke_provider_fn::__rust_begin_short_backtrace
  22:     0x7f1dd054a729 - rustc_query_impl[363d8420d406e095]::execution::try_execute_query::<rustc_data_structures[2dcf1c254b064446]::vec_cache::VecCache<rustc_span[5521c16fdf99d052]::def_id::LocalDefId, rustc_middle[eddce6fccf027369]::query::erase::ErasedData<[u8; 8usize]>, rustc_middle[eddce6fccf027369]::dep_graph::graph::DepNodeIndex>, false>
  23:     0x7f1dd054a183 - rustc_query_impl[363d8420d406e095]::query_impl::typeck_root::execute_query_non_incr::__rust_end_short_backtrace
  24:     0x7f1dd0542a80 - rustc_mir_build[9f2b81bfde8b0892]::thir::pattern::check_match::check_match
  25:     0x7f1dd0541e07 - rustc_query_impl[363d8420d406e095]::query_impl::check_match::invoke_provider_fn::__rust_begin_short_backtrace
  26:     0x7f1dd106b4f1 - rustc_query_impl[363d8420d406e095]::execution::try_execute_query::<rustc_data_structures[2dcf1c254b064446]::vec_cache::VecCache<rustc_span[5521c16fdf99d052]::def_id::LocalDefId, rustc_middle[eddce6fccf027369]::query::erase::ErasedData<[u8; 1usize]>, rustc_middle[eddce6fccf027369]::dep_graph::graph::DepNodeIndex>, false>
  27:     0x7f1dd106b1f6 - rustc_query_impl[363d8420d406e095]::query_impl::check_match::execute_query_non_incr::__rust_end_short_backtrace
  28:     0x7f1dd0553e93 - rustc_mir_build[9f2b81bfde8b0892]::builder::build_mir_inner_impl
  29:     0x7f1dd020fb6a - rustc_mir_transform[3af5eea067a9e4c2]::mir_built
  30:     0x7f1dd054a729 - rustc_query_impl[363d8420d406e095]::execution::try_execute_query::<rustc_data_structures[2dcf1c254b064446]::vec_cache::VecCache<rustc_span[5521c16fdf99d052]::def_id::LocalDefId, rustc_middle[eddce6fccf027369]::query::erase::ErasedData<[u8; 8usize]>, rustc_middle[eddce6fccf027369]::dep_graph::graph::DepNodeIndex>, false>
  31:     0x7f1dd054a3c3 - rustc_query_impl[363d8420d406e095]::query_impl::mir_built::execute_query_non_incr::__rust_end_short_backtrace
  32:     0x7f1dd023029d - rustc_mir_transform[3af5eea067a9e4c2]::ffi_unwind_calls::has_ffi_unwind_calls
  33:     0x7f1dd022fcd9 - rustc_query_impl[363d8420d406e095]::query_impl::has_ffi_unwind_calls::invoke_provider_fn::__rust_begin_short_backtrace
  34:     0x7f1dd106b4f1 - rustc_query_impl[363d8420d406e095]::execution::try_execute_query::<rustc_data_structures[2dcf1c254b064446]::vec_cache::VecCache<rustc_span[5521c16fdf99d052]::def_id::LocalDefId, rustc_middle[eddce6fccf027369]::query::erase::ErasedData<[u8; 1usize]>, rustc_middle[eddce6fccf027369]::dep_graph::graph::DepNodeIndex>, false>
  35:     0x7f1dd106b0c2 - rustc_query_impl[363d8420d406e095]::query_impl::has_ffi_unwind_calls::execute_query_non_incr::__rust_end_short_backtrace
  36:     0x7f1dd0d2910a - rustc_mir_transform[3af5eea067a9e4c2]::mir_promoted
  37:     0x7f1dd0d28f6e - rustc_query_impl[363d8420d406e095]::query_impl::mir_promoted::invoke_provider_fn::__rust_begin_short_backtrace
  38:     0x7f1dd0575c81 - rustc_query_impl[363d8420d406e095]::execution::try_execute_query::<rustc_data_structures[2dcf1c254b064446]::vec_cache::VecCache<rustc_span[5521c16fdf99d052]::def_id::LocalDefId, rustc_middle[eddce6fccf027369]::query::erase::ErasedData<[u8; 16usize]>, rustc_middle[eddce6fccf027369]::dep_graph::graph::DepNodeIndex>, false>
  39:     0x7f1dd104b3c3 - rustc_query_impl[363d8420d406e095]::query_impl::mir_promoted::execute_query_non_incr::__rust_end_short_backtrace
  40:     0x7f1dd104b5a9 - rustc_borrowck[7b06359707d7b718]::mir_borrowck
  41:     0x7f1dd104b4a5 - rustc_query_impl[363d8420d406e095]::query_impl::mir_borrowck::invoke_provider_fn::__rust_begin_short_backtrace
  42:     0x7f1dd054a729 - rustc_query_impl[363d8420d406e095]::execution::try_execute_query::<rustc_data_structures[2dcf1c254b064446]::vec_cache::VecCache<rustc_span[5521c16fdf99d052]::def_id::LocalDefId, rustc_middle[eddce6fccf027369]::query::erase::ErasedData<[u8; 8usize]>, rustc_middle[eddce6fccf027369]::dep_graph::graph::DepNodeIndex>, false>
  43:     0x7f1dd054c357 - rustc_query_impl[363d8420d406e095]::query_impl::mir_borrowck::execute_query_non_incr::__rust_end_short_backtrace
  44:     0x7f1dcec416f2 - <rustc_hir_analysis[4b6d32d55a266fd3]::collect::type_of::opaque::TaitConstraintLocator>::check
  45:     0x7f1dcec80601 - rustc_hir_analysis[4b6d32d55a266fd3]::collect::type_of::opaque::find_opaque_ty_constraints_for_tait
  46:     0x7f1dd2016111 - rustc_hir_analysis[4b6d32d55a266fd3]::collect::type_of::type_of_opaque.cold
  47:     0x7f1dd025298d - rustc_query_impl[363d8420d406e095]::execution::try_execute_query::<rustc_middle[eddce6fccf027369]::query::caches::DefIdCache<rustc_middle[eddce6fccf027369]::query::erase::ErasedData<[u8; 8usize]>>, false>
  48:     0x7f1dd13b111a - rustc_query_impl[363d8420d406e095]::query_impl::type_of_opaque::execute_query_non_incr::__rust_end_short_backtrace
  49:     0x7f1dd0524a3d - rustc_hir_analysis[4b6d32d55a266fd3]::collect::type_of::type_of
  50:     0x7f1dd05182a6 - rustc_query_impl[363d8420d406e095]::query_impl::type_of::invoke_provider_fn::__rust_begin_short_backtrace
  51:     0x7f1dd025298d - rustc_query_impl[363d8420d406e095]::execution::try_execute_query::<rustc_middle[eddce6fccf027369]::query::caches::DefIdCache<rustc_middle[eddce6fccf027369]::query::erase::ErasedData<[u8; 8usize]>>, false>
  52:     0x7f1dd0252510 - rustc_query_impl[363d8420d406e095]::query_impl::type_of::execute_query_non_incr::__rust_end_short_backtrace
  53:     0x7f1dd0fe268f - rustc_hir_analysis[4b6d32d55a266fd3]::check::check::check_opaque
  54:     0x7f1dd108a5e3 - rustc_hir_analysis[4b6d32d55a266fd3]::check::check::check_item_type
  55:     0x7f1dd106bf30 - rustc_hir_analysis[4b6d32d55a266fd3]::check::wfcheck::check_well_formed
  56:     0x7f1dd106bf11 - rustc_query_impl[363d8420d406e095]::query_impl::check_well_formed::invoke_provider_fn::__rust_begin_short_backtrace
  57:     0x7f1dd106b4f1 - rustc_query_impl[363d8420d406e095]::execution::try_execute_query::<rustc_data_structures[2dcf1c254b064446]::vec_cache::VecCache<rustc_span[5521c16fdf99d052]::def_id::LocalDefId, rustc_middle[eddce6fccf027369]::query::erase::ErasedData<[u8; 1usize]>, rustc_middle[eddce6fccf027369]::dep_graph::graph::DepNodeIndex>, false>
  58:     0x7f1dd106b287 - rustc_query_impl[363d8420d406e095]::query_impl::check_well_formed::execute_query_non_incr::__rust_end_short_backtrace
  59:     0x7f1dd1067d8e - rustc_hir_analysis[4b6d32d55a266fd3]::check::wfcheck::check_type_wf
  60:     0x7f1dd1067c13 - rustc_query_impl[363d8420d406e095]::query_impl::check_type_wf::invoke_provider_fn::__rust_begin_short_backtrace
  61:     0x7f1dd13a7446 - rustc_query_impl[363d8420d406e095]::execution::try_execute_query::<rustc_middle[eddce6fccf027369]::query::caches::SingleCache<rustc_middle[eddce6fccf027369]::query::erase::ErasedData<[u8; 1usize]>>, false>
  62:     0x7f1dd13a7229 - rustc_query_impl[363d8420d406e095]::query_impl::check_type_wf::execute_query_non_incr::__rust_end_short_backtrace
  63:     0x7f1dd05476da - rustc_hir_analysis[4b6d32d55a266fd3]::check_crate
  64:     0x7f1dd054b43a - rustc_interface[b2016dfbaec51b1d]::passes::analysis
  65:     0x7f1dd13a8247 - rustc_query_impl[363d8420d406e095]::execution::try_execute_query::<rustc_middle[eddce6fccf027369]::query::caches::SingleCache<rustc_middle[eddce6fccf027369]::query::erase::ErasedData<[u8; 0usize]>>, false>
  66:     0x7f1dd13a7f29 - rustc_query_impl[363d8420d406e095]::query_impl::analysis::execute_query_non_incr::__rust_end_short_backtrace
  67:     0x7f1dd137c0c0 - rustc_interface[b2016dfbaec51b1d]::interface::run_compiler::<(), rustc_driver_impl[5bc58d79d7fe6386]::run_compiler::{closure#0}>::{closure#1}
  68:     0x7f1dd13bcb3a - std[fd5358557e4939aa]::sys::backtrace::__rust_begin_short_backtrace::<rustc_interface[b2016dfbaec51b1d]::util::run_in_thread_with_globals<rustc_interface[b2016dfbaec51b1d]::util::run_in_thread_pool_with_globals<rustc_interface[b2016dfbaec51b1d]::interface::run_compiler<(), rustc_driver_impl[5bc58d79d7fe6386]::run_compiler::{closure#0}>::{closure#1}, ()>::{closure#0}, ()>::{closure#0}::{closure#0}, ()>
  69:     0x7f1dd13bc8ed - <std[fd5358557e4939aa]::thread::lifecycle::spawn_unchecked<rustc_interface[b2016dfbaec51b1d]::util::run_in_thread_with_globals<rustc_interface[b2016dfbaec51b1d]::util::run_in_thread_pool_with_globals<rustc_interface[b2016dfbaec51b1d]::interface::run_compiler<(), rustc_driver_impl[5bc58d79d7fe6386]::run_compiler::{closure#0}>::{closure#1}, ()>::{closure#0}, ()>::{closure#0}::{closure#0}, ()>::{closure#1} as core[824893b31d0d62cd]::ops::function::FnOnce<()>>::call_once::{shim:vtable#0}
  70:     0x7f1dd13b2f90 - <std[fd5358557e4939aa]::sys::thread::unix::Thread>::new::thread_start
  71:     0x7f1dcac981b9 - <unknown>
  72:     0x7f1dcad1d21c - <unknown>
  73:                0x0 - <unknown>

error: the compiler unexpectedly panicked. This is a bug

note: we would appreciate a bug report: https://github.com/rust-lang/rust/issues/new?labels=C-bug%2C+I-ICE%2C+T-compiler&template=ice.md

note: please make sure that you have updated to the latest nightly

note: rustc 1.98.0-nightly (fd07dbfc9 2026-06-28) running on x86_64-unknown-linux-gnu

note: compiler flags: -Z next-solver=globally -Z dump-mir-dir=dir

query stack during panic:
#0 [typeck_root] type-checking `bar`
#1 [check_match] match-checking `bar`
#2 [mir_built] building MIR for `bar`
#3 [has_ffi_unwind_calls] checking if `bar` contains FFI-unwind calls
#4 [mir_promoted] promoting constants in MIR for `bar`
#5 [mir_borrowck] borrow-checking `bar`
#6 [type_of_opaque] computing type of opaque `Tait::{opaque#0}`
#7 [type_of] computing type of `Tait::{opaque#0}`
#8 [check_well_formed] checking that `Tait::{opaque#0}` is well-formed
#9 [check_type_wf] checking that types are well-formed
#10 [analysis] running analysis passes on crate `mvce`
end of query stack
error: aborting due to 4 previous errors; 1 warning emitted

Some errors have detailed explanations: E0277, E0601, E0658.
For more information about an error, try `rustc --explain E0277`.

```

</p>
</details>

<!--
query stack:
436 | pub struct Vec<T, #[unstable(feature = "allocator_api", issue = "32838")] A: Allocator = Global> {
#0 [typeck_root] type-checking `bar`
#1 [check_match] match-checking `bar`
#2 [mir_built] building MIR for `bar`
#3 [has_ffi_unwind_calls] checking if `bar` contains FFI-unwind calls
#4 [mir_promoted] promoting constants in MIR for `bar`
#5 [mir_borrowck] borrow-checking `bar`
#6 [type_of_opaque] computing type of opaque `Tait::{opaque#0}`
#7 [type_of] computing type of `Tait::{opaque#0}`
#8 [check_well_formed] checking that `Tait::{opaque#0}` is well-formed
#9 [check_type_wf] checking that types are well-formed
#10 [analysis] running analysis passes on crate `mvce`

-->
@rustbot label +F-inherent_associated_types +F-type_alias_impl_trait +WG-trait-system-refactor

--- a/compiler/rustc_hir_analysis/src/hir_ty_lowering/errors.rs
+++ b/compiler/rustc_hir_analysis/src/hir_ty_lowering/errors.rs
@@ -984,15 +984,17 @@ impl<'tcx> dyn HirTyLowerer<'tcx> + '_ {
                 ty::PredicateKind::Clause(ty::ClauseKind::Projection(pred)) => {
                     // `<Foo as Iterator>::Item = String`.
                     let projection_term = pred.projection_term;
+                    let term = pred.term;
+                    let self_ty = projection_term.args.get(0).and_then(|arg| arg.as_type())?;
+
+                    let obligation = format!("{projection_term} = {term}");
                     let quiet_projection_term = projection_term
                         .with_replaced_self_ty(tcx, Ty::new_var(tcx, ty::TyVid::ZERO));
-
-                    let term = pred.term;
-                    let obligation = format!("{projection_term} = {term}");
                     let quiet = format!("{quiet_projection_term} = {term}");
 
-                    bound_span_label(projection_term.self_ty(), &obligation, &quiet);
-                    Some((obligation, projection_term.self_ty()))
+                    bound_span_label(self_ty, &obligation, &quiet);
+
+                    Some(obligation)
                 }
                 ty::PredicateKind::Clause(ty::ClauseKind::Trait(poly_trait_ref)) => {
                     let p = poly_trait_ref.trait_ref;
@@ -1001,7 +1003,7 @@ impl<'tcx> dyn HirTyLowerer<'tcx> + '_ {
                     let obligation = format!("{self_ty}: {path}");
                     let quiet = format!("_: {path}");
                     bound_span_label(self_ty, &obligation, &quiet);
-                    Some((obligation, self_ty))
+                    Some(obligation)
                 }
                 _ => None,
             }
@@ -1013,7 +1015,7 @@ impl<'tcx> dyn HirTyLowerer<'tcx> + '_ {
             .into_iter()
             .map(|error| error.root_obligation.predicate)
             .filter_map(format_pred)
-            .map(|(p, _)| format!("`{p}`"))
+            .map(|p| format!("`{p}`"))
             .collect();
         bounds.sort();
         bounds.dedup();

--- a/tests/ui/associated-inherent-types/next-solver-unsatisfied-bounds-inherent-tait.rs
+++ b/tests/ui/associated-inherent-types/next-solver-unsatisfied-bounds-inherent-tait.rs
@@ -1,0 +1,22 @@
+//@ compile-flags: -Znext-solver=globally
+
+#![feature(inherent_associated_types)]
+#![feature(type_alias_impl_trait)]
+#![allow(incomplete_features)]
+
+struct Foo<T>(T);
+
+impl Foo<Vec<[u32]>> {
+    //~^ ERROR the size for values of type `[u32]` cannot be known at compilation time
+    type Assoc = u32;
+}
+
+type Tait = impl Sized;
+
+#[define_opaque(Tait)]
+fn bar() {
+    let _: Foo<Tait>::Assoc = 42;
+    //~^ ERROR the associated type `Assoc` exists for `Foo<Tait>`, but its trait bounds were not satisfied
+}
+
+fn main() {}

--- a/tests/ui/associated-inherent-types/next-solver-unsatisfied-bounds-inherent-tait.stderr
+++ b/tests/ui/associated-inherent-types/next-solver-unsatisfied-bounds-inherent-tait.stderr
@@ -1,0 +1,22 @@
+error[E0277]: the size for values of type `[u32]` cannot be known at compilation time
+  --> $DIR/next-solver-unsatisfied-bounds-inherent-tait.rs:9:10
+   |
+LL | impl Foo<Vec<[u32]>> {
+   |          ^^^^^^^^^^ doesn't have a size known at compile-time
+   |
+   = help: the trait `Sized` is not implemented for `[u32]`
+note: required by an implicit `Sized` bound in `Vec`
+  --> $SRC_DIR/alloc/src/vec/mod.rs:LL:COL
+
+error: the associated type `Assoc` exists for `Foo<Tait>`, but its trait bounds were not satisfied
+  --> $DIR/next-solver-unsatisfied-bounds-inherent-tait.rs:18:23
+   |
+LL | struct Foo<T>(T);
+   | ------------- associated type `Assoc` not found for this struct
+...
+LL |     let _: Foo<Tait>::Assoc = 42;
+   |                       ^^^^^ associated type cannot be referenced on `Foo<Tait>` due to unsatisfied trait bounds
+
+error: aborting due to 2 previous errors
+
+For more information about this error, try `rustc --explain E0277`.

